File: Sem4/MPP_Templates/ZboruriCSharp/networking/FlightClientObjectWorker.cs
using model;
using services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace networking
{
    public class FlightClientObjectWorker : IFlightObserver
    {
        private IFlightServer server;
        private TcpClient connection;

        private NetworkStream stream;
        private IFormatter formatter;
        private volatile bool connected;

        public FlightClientObjectWorker(IFlightServer server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void run()
        {
            while (connected)
            {
                try
                {
                    object request = formatter.Deserialize(stream);
                    object response = handleRequest((Request)request);
                    if (response != null)
                    {
                        sendResponse((Response)response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }
        private void sendResponse(Response response)
        {
            Console.WriteLine("Sending response " + response);
            formatter.Serialize(stream, response);
            stream.Flush();
        }

        private object handleRequest(Request request)
        {
            Response response = null;

            if (request is LoginRequest)
            {
                Console.WriteLine("Login request ...");
                LoginRequest logReq = (LoginRequest)request;
                Angajat angajat = logReq.Angajat;
                try
                {
                    lock (server)
                    {
                        server.LogIn(angajat, this);
                    }
                    return new OkResponse();
                }
                catch (FlightException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }


            if (request is LogoutRequest)
            {
                Console.WriteLine("Logout request ...");
                LogoutRequest logReq = (LogoutRequest)request;
                Angajat angajat = logReq.User;
                try
                {
                    lock (server)
                    {
                        server.LogOut(angajat, this);
                    }
                    return new OkResponse();
                }
                catch (FlightException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is GetZboruriRequest)
            {
                Console.WriteLine("Flights get request");
                GetZboruriRequest getReq = (GetZboruriRequest)request;
                IEnumerable<Zbor> zboruri;
                try
                {
                    lock (server)
                    {
                        zboruri = server.GetZboruri();
                    }
                    return new GetZboruriResponse(zboruri);
                }
                catch (FlightException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is GetZboruriCautareRequest)
            {
                Console.WriteLine("Find fligts request");
                GetZboruriCautareRequest getReq = (GetZboruriCautareRequest)request;
                ZborDTO2 zbor = getReq.zborDTO2;
                IEnumerable<ZborDTO> zborDTOs;
                try
                {
                    lock (server)
                    {
                        zborDTOs = server.GetZboruriCautare(zbor.destinatie, zbor.data);
                    }
                    return new GetZboruriCautareResponse(zborDTOs);
                }
                catch (FlightException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is FindClientRequest)
            {
                Console.WriteLine("Finding client");
                FindClientRequest findReq = (FindClientRequest)request;
                ClientDTO clientDTO = findReq.Client;
                Client client;
                try
                {
                    lock (server)
                    {
                        client = server.FindClient(clientDTO.nume, clientDTO.adresa);
                    }
                    return new FindClientResponse(client);
                }
                catch (FlightException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is FindZborRequest)
            {
                Console.WriteLine("Finding flight");
                FindZborRequest findReq = (FindZborRequest)request;
                ZborDTO3 zborDTO3 = findReq.ZborDTO3;
                Zbor zbor;
                try
                {
                    lock (server)
                    {
                        zbor = server.FindZbor(zborDTO3.destinatie, zborDTO3.data, zborDTO3.ora);
                    }
                    return new FindZborResponse(zbor);
                }
                catch (FlightException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is FindIndexRequest)
            {
                Console.WriteLine("Finding index");
                FindIndexRequest findReq = (FindIndexRequest)request;
                int index;
                try
                {
                    lock (server)
                    {
                        index = server.FindIndex();
                    }
                    return new FindIndexResponse(index);
                }
                catch (FlightException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is BuyTickeRequest)
            {
                Console.WriteLine("Ticket bought");
                BuyTickeRequest req = (BuyTickeRequest)request;
                Bilet bilet = req.Bilet;
                int id = req.Id;
                int nr = req.NrLocuri;
                try
                {
                    lock (server)
                    {
                        server.CumparaBilet(bilet, id, nr);
                    }
                    return new OkResponse();
                }catch(FlightException e)
                {
                    Console.WriteLine(e.Message);
                }
            }
            return response;
        }

        public void TicketBought(IEnumerable<Zbor> zboruri)
        {
            try
            {
                sendResponse(new BuyTicketResponse(zboruri));
            }
            catch (FlightException e)
            {
                Console.WriteLine(e.Message);
            }
        }
    }
}

File: Sem4/MPP_Templates/Zboruri/FlightNetworking/src/main/java/objectprotocol/FlightClientObjectWorker.java
package objectprotocol;

import flight.domain.*;
import dto.*;
import flight.domain.*;
import services.FlightException;
import services.IFlightObserver;
import services.IFlightServer;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;

public class FlightClientObjectWorker implements Runnable, IFlightObserver {
    private IFlightServer server;
    private Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public FlightClientObjectWorker(IFlightServer server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try{
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        }catch(IOException e){
            e.printStackTrace();
        }
    }

    public void run(){
        while(connected){
            try{
                Object request = input.readObject();
                Object response = handleRequest((Request) request);
                if(response!=null){
                    sendResponse((Response) response);
                }
            }catch(IOException e){
                e.printStackTrace();
            }catch (ClassNotFoundException e){
                e.printStackTrace();
            } catch (FlightException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    private void sendResponse(Response response) throws IOException{
        System.out.println("Sending response " + response);
        output.writeObject(response);
        output.flush();
    }

    private Response handleRequest(Request request) throws FlightException {
        Response response = null;
        if(request instanceof LoginRequest){
            System.out.println("Login request ...");
            LoginRequest logReq = (LoginRequest) request;
            Angajat angajat = logReq.getUser();
            try{
                server.logIn(angajat, this);
                return new OkResponse();
            } catch (FlightException e) {
                connected = false;
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof LogoutRequest){
            System.out.println("LogoutRequest");
            LogoutRequest logReq = (LogoutRequest) request;
            Angajat angajat = logReq.getUser();
            try{
                server.logOut(angajat,this);
                connected = false;
                return new OkResponse();
            }catch (FlightException e){
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof GetZboruriRequest){
            System.out.println("GetZboruriRequest");
            try{
                Iterable<Zbor> zboruri = server.getZboruri();
                return new GetZboruriResponse(zboruri);
            } catch (FlightException e) {
                connected = false;
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof GetZboruriCautateRequest){
            GetZboruriCautateRequest getZboruriCautateRequest = (GetZboruriCautateRequest) request;
            ZborDTO2 zborDTO2 = getZboruriCautateRequest.getZborDTO2();
            System.out.println("GetZboruriCautateRequest");
            try{
                Iterable<ZborDTO> zboruri = server.getZboruriCautare(zborDTO2.getDestinatie(),zborDTO2.getPlecare());
                return new GetZboruriCautateResponse(zboruri);
            } catch (FlightException e) {
                connected = false;
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof FindClientRequest){
            FindClientRequest findClientRequest = (FindClientRequest) request;
            ClientDTO client = findClientRequest.getClient();
            System.out.println("FindClientRequest");
            try{
                Client client1 = server.findClient(client.getNume(),client.getAdresa());
                return new FindClientResponse(client1);
            } catch (FlightException e) {
               return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof FindFlightRequest){
            FindFlightRequest findFlightRequest = (FindFlightRequest) request;
            ZborDTO3 zborDTO3 = findFlightRequest.getZborDTO3();
            System.out.println("FindFlightRequest");
            try{
                Zbor zbor = server.findZbor(zborDTO3.getDestinatie(),zborDTO3.getData(),zborDTO3.getOra());
                return new FindFlightResponse(zbor);
            } catch (FlightException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof FindIndexRequest){
            System.out.println("FindIndexRequest");
            try{
                Integer index = server.findIndex();
                return new FindIndexResponse(new IndexDTO(index));
            } catch (FlightException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof BuyTicketRequest){
            BuyTicketRequest buyTicketRequest = (BuyTicketRequest) request;
            Bilet bilet = buyTicketRequest.getBilet();
            int id = buyTicketRequest.getId();
            System.out.println("BuyTicketRequest");
            server.cumparaBilet(bilet, id);
            return new OkResponse();
        }

        if (request instanceof ZborUpdateRequest){
            ZborUpdateRequest zborUpdateRequest = (ZborUpdateRequest) request;
            ZborDTO4 zborDTO4 = zborUpdateRequest.getZborDTO4();
            System.out.println("ZborUpdateRequest");
            try{
                server.zborUpdate(zborDTO4.getZborID(),zborDTO4.getLocuri());
                return new ZborUpdateResponse();
            } catch (FlightException e) {
                return new ErrorResponse(e.getMessage());
            }
        }
        return response;
    }

    @Override
    public void ticketBought(Iterable<Zbor> zboruri) {
        System.out.println("Ticket has been bought");
        try {
            sendResponse(new BuyTicketResponse(zboruri));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
